### DISCUSSION TITLE
Proposal: Migrating homebrew-fonts to homebrew-cask-fonts

### DISCUSSION THREAD

**Discussion by tani** [Homebrew/discussions#2316] [Linux] [answered]
# Motivation

I am a maintainer of [linuxbrew/homebrew-fonts](https://github.com/linuxbrew/homebrew-fonts). I got a proposal by @MikeMcQuaid proposed changing the place of homebrew-fonts (https://github.com/Linuxbrew/homebrew-fonts/issues/28). I propose merging it to homebrew-cask-fonts. I dream that we install the same fonts from the same repository by the same instruction between macOS and Linux like Homebrew.

# Problems

First, I explain the four problems.
- Philosophical problem. Linux support is out of scope for this project goal.
- Technical problem. Cask is designed only for macOS.
- Resource problem. Cask maintainers have no time to maintain the extra packages.
- Legal problem. Some cask is only for macOS.

## Philosophical problem

A few years ago, Homebrew was a project only for macOS, but they changed their targets. They started to support Linux packages.
This decision is an excellent contribution to the equality of packages between macOS and Linux. As the same as this case, changing the philosophy makes this repository a perfect project. Equality is essential for open source, isn't it?

## Technical problem

Generally speaking, porting a package from macOS to Liunx is tricky because a machine architecture, a binary format, and everything are different. The early version of cask treated with dmg package, which is a special format for macOS and contains binary files run on the macOS. This case is the most difficult situation to port to Linux. However, porting font is not so hard because this is a zip file or tarball. Furthermore, the binary format is standardized in TTF and OTF. For our purpose, changing a font location is enough to port.  Therefore, there is no technical problem.
 
## Resource problem

I know the resource is limited. Thus, We need to focus on one target.
Hence, I do not propose merging and maintaining homebrew-fonts to this repository without my help. I can maintain the Linux port. All steps to sync the latest casks are fully automated. So, in fact, I have been keeping homebrew-fonts for four years. 

## Legal problem

This is perhaps the most serious problem of the four problems. Most of the fonts are licensed under the OSS Licenses, but a few fonts are not.
For example, San Francisco fonts are only for macOS. Thus, we have to take care of the license of each font for Linux.

# Approach

I have a script to convert a cask file to a formula. I am running this script daily.

# Note

Four years ago, I was impressed with this repository for convenience and launched a fork named linuxbrew-fonts. Thanks to the many contributions, I could keep the latest formulas, and the contributors told me the importance of equality to install the package as the same as macOS.

Two years ago, I was excited to merge the linuxbrew formulas to homebrew formulas. The equality of packages is more important than before. I propose connecting my work to your repository for equality with my respect to your work.

**Comment by MikeMcQuaid** [comment]
Thanks for the write-up @tani! I'm interested in @Homebrew/cask and particularly @vitorgalvao's thoughts here.

One thought of my own when reading this: what's the advantage of a Cask for fonts over a Formula? Are the Casks e.g. installed into my macOS fonts whereas a Formula would require a manual intervention?

**Comment by vitorgalvao** [comment]
> what's the advantage of a Cask for fonts over a Formula?

If formulae can unpack files directly to `~/Library/Fonts`, there’s no technical advantage. I’d welcome it. It they can’t be unpacked there but can be symlinked, that would also work (assuming macOS hasn’t broken it since the time we stopped doing it that way in casks). I remember well that fonts were introduced as casks as a sort of “why not, it’s just another artefact which can be implemented as an extension of what we already support”. It was well before HB and HBC merged.

> Legal problem

In casks we had support for licenses (it was specifically was introduced because of fonts). I disliked that because it was the number one cause of held-up PRs. It caused a bunch of extra effort and an increased backlog for zero benefit. We eventually got rid of it. We only download the fonts; I don’t necessarily agree it should be our responsibility to read and interpret their licenses (they change depending on context, e.g. personal VS commercial use). But I’m not opposed to licenses being there, I just don’t want to subject myself to that problem again.

> I propose connecting my work to your repository for equality with my respect to your work.

These days, for a new font to be accepted it has to be exceptionally popular. All others are automatically handled by converting from the Google Fonts repo. See [CONTRIBUTING](https://github.com/Homebrew/homebrew-cask-fonts/blob/master/CONTRIBUTING.md#when-to-contribute) for details. I should be possible to add your automations on top of that.

Assuming fonts could be formulae on macOS, we could convert everything to fonts and use the same files for macOS and Linux, just changing the installation/symlinking directory depending on OS. Then slowly readd licenses.

**Reply by MikeMcQuaid** [reply]
> It they can’t be unpacked there but can be symlinked, that would also work (assuming macOS hasn’t broken it since the time we stopped doing it that way in casks).

Can't be unpacked there but could be symlinked (although, as-is, not automatically but perhaps with some `brew fonts` helper tool).

> I don’t necessarily agree it should be our responsibility to read and interpret their licenses (they change depending on context, e.g. personal VS commercial use).

Agreed.

**Comment by tani** [comment, marked as answer]
Thank you for your generous replys, @MikeMcQuaid and @vitorgalvao .

> In casks we had support for licenses (it was specifically was introduced because of fonts). I disliked that because it was the number one cause of held-up PRs. It caused a bunch of extra effort and an increased backlog for zero benefit. We eventually got rid of it. We only download the fonts; I don’t necessarily agree it should be our responsibility to read and interpret their licenses (they change depending on context, e.g. personal VS commercial use). But I’m not opposed to licenses being there, I just don’t want to subject myself to that problem again.

I understand your opinion. Respecting the license, e.g., personal VS commercial use, is user responsibility and not maintainers'. 

> All others are automatically handled by converting from the Google Fonts repo.

Great. I will create a PR modifying a template to generate formulae simultaneously.

> See CONTRIBUTING for details. I should be possible to add your automations on top of that.

Okay. I will send several PRs to run my script on your workflows. I suspect this migration takes a bit of time.

Phase 1: Updating a script to generate formulae from Google Fonts
Phase 2: Adding a new script to convert casks excepting Google Fonts to formulae
Phase 3: Integration of current workflow using the above two scripts

Then, the homebrew-cask-fonts contains two kinds of packages, i.e., formulae and casks.

> Assuming fonts could be formulae on macOS, we could convert everything to fonts and use the same files for macOS and Linux, just changing the installation/symlinking directory depending on OS. Then slowly readd licenses.

After the integration, we can discuss the suitable package format for fonts like symlinking vs copying.

If anyone has any consideration, tell me, please.

**Reply by MikeMcQuaid** [reply]
@tani Hold off any more work for a couple of days while we give other maintainers a chance to chime in.

> Then, the homebrew-cask-fonts contains two kinds of packages, i.e., formulae and casks.

I think it would probably make sense to have only formulae?




> After the integration, we can discuss the suitable package format for fonts like symlinking vs copying.

Probably want a `brew fonts` external command that lives in the homebrew/cask-fonts tap (like `https://github.com/Homebrew/homebrew-core/blob/master/cmd/aspell-dictionaries.rb`) and consider renaming the tap to `homebrew/fonts`.

**Reply by tani** [reply]
I see. I am looking forward to seeing other maintainers feedback. Having only formulae is a nice idea but we need to announce the migration to the existing users as you know. Thus, I prefer the graceful migration; First, we publish two kinds of package formats and next we integrate them into one package format. I also would like to hear other comments for your idea of adding an external command.

**Reply by Rylan12** [reply]
I have a few thoughts, mainly because from what I understand it seems that users who currently have cask fonts installed are going to suddenly be hit with a confusing state where their current fonts can't be upgraded and they need to switch taps and reinstall everything.

> Probably want a `brew fonts` external command that lives in the homebrew/cask-fonts tap

@MikeMcQuaid can you elaborate on what exactly this command would do?

---

From my understanding it sounds like the best situation is for all fonts to become formulae so they can be installed on either OS (and, if possible, the tap can be renamed to `homebrew/fonts`). However, this would require a few things:

1. A way for existing cask installs to be replaced with the corresponding formula without needing to manually uninstall the cask and install the formula
2. A way for the tap rename to automatically happen (i.e. running `brew update` replaces `homebrew/cask-fonts` with `homebrew/fonts`)

From what I know, those aren't currently possible. One idea for how we could accomplish this is to add a migration command (maybe this is the `brew fonts` command that you were talking about) to `homebrew/cask-fonts`. Personally, I'd call this `brew migrate-fonts`. This would be a one-time-use command that would (possibly not in this order):

1. Uninstall fonts from `homebrew/cask-fonts` that are currently installed (and remember their names)
2. Untap `homebrew/cask-fonts`
3. Tap `homebrew/fonts`
4. Install the `homebrew/fonts` font formulae for those that were uninstalled previously 

Also, we would need to have some sort of message (maybe it's something in `brew update`, `brew.sh`, `brew doctor`, or `brew install` on a font in `homebrew/cask-fonts`) that instructs the user to run this command.

While this, of course, it not as seamless as all of that happening automatically, I think it might be a good middle ground because we probably don't want to add all this functionality for one migration if that makes sense (i.e. we don't usually need to support migrating a cask for a formula).

**Reply by MikeMcQuaid** [reply]
> I have a few thoughts, mainly because from what I understand it seems that users who currently have cask fonts installed are going to suddenly be hit with a confusing state where their current fonts can't be upgraded and they need to switch taps and reinstall everything.

We have an existing migration path for casks to formulae
- https://github.com/Homebrew/homebrew-cask/blob/master/tap_migrations.json#L2
- https://github.com/Homebrew/brew/blob/033c365b0da986efd8e5b7e14111e19afa6bdccf/Library/Homebrew/cmd/update-report.rb#L416-L438

> @MikeMcQuaid can you elaborate on what exactly this command would do?

Like `brew services`: do thinks that the Homebrew sandbox of formulae won't normally allow. In this case, symlink or copy to `~/Library/Fonts` (and whatever the Linux equivalent is)

**Reply by Rylan12** [reply]
> We have an existing migration path for casks to formulae

Ah, thanks for the link! I didn't realize we had that. That's cool!

> Like `brew services`: do thinks that the Homebrew sandbox of formulae won't normally allow. In this case, symlink or copy to `~/Library/Fonts` (and whatever the Linux equivalent is)

👍

**Reply by carlocab** [reply]
I'd prefer, if possible, to leave fonts as casks.

Casks are already designed to install and uninstall files in arbitrary locations on the file system. (At the very least, they are much better suited for doing this than formulae are.) I think that makes them better for fonts, which aren't really meant to live in `HOMEBREW_PREFIX` (except perhaps for the purposes of keeping track of them for uninstallation later).

This might also serve as a nice opportunity to lay the groundwork for making Homebrew Cask cross-platform (if we ever want to consider doing this at some point -- not that it would obligate us to).

**Reply by vitorgalvao** [reply]
> This might also serve as a nice opportunity to lay the groundwork for making Homebrew Cask cross-platform (if we ever want to consider doing this at some point -- not that it would obligate us to).

Historically, we’ve always been actively opposed to it because dealing with macOS’ myriad of software types and developer inconsistency is hard enough as it is.

**Reply by MikeMcQuaid** [reply]
> This might also serve as a nice opportunity to lay the groundwork for making Homebrew Cask cross-platform (if we ever want to consider doing this at some point -- not that it would obligate us to).

This would be an enormous amount of work just for "casks are already designed for this". I agree with @vitorgalvao that I don't think it makes any sense to do so.

---

The feasible options I see it are:
1. migrate homebrew-cask-fonts formulae to tani's personal GitHub account for Linux-only
2. migrate homebrew-cask-fonts to homebrew-fonts which can be shared between macOS and Linux and managed in the Homebrew organisation and provide a `brew fonts` to do any of the symlinking required on macOS

We're also trying to reduce the number of Cask taps we maintain and my understanding is that there has been some issues with the homebrew-cask-fonts repository and e.g. Subversion (@vitorgalvao can probably add more information) that may make migration to formulae superior for folks on macOS, too.

**Reply by carlocab** [reply]
> This would be an enormous amount of work just for "casks are already designed for this".

Per @colindean on Slack, it seems that making font casks work on Linux requires changing only a handful of lines of code.

**Reply by colindean** [reply]
POC, makes `brew install font-libre-franklin font-gwendolyn` work, the latter only if you have subversion install obviously: 

https://github.com/Homebrew/brew/compare/master...colindean:allow-font-casks-on-linux

Font casks merely move files, so as long as the user has set
```bash
HOMEBREW_CASK_OPTS=--fontdir=${HOME}/.fonts
```
on Linux, they can install fonts to their home directory. I did this, things work. Uninstall also works.

Another future possibility would be to install these somewhere in the `HOMEBREW_PREFIX` and instruct the user to symlink _into_ that directory, e.g. `ln -s $(brew --prefix)/share/fonts ${HOME}/.fonts/homebrew`.

**Reply by tani** [reply]
This is awesome information. This is the best way to complete my proposal.
However, my concern is that the target of the cask becomes unclear because
it enables casks for Linux if the cask has only fonts compared to disabling Linux completely at this moment.
This condition is a bit unclear for users. Anyway, I am glad to move forward with this proposal with your idea.

**Reply by sjackman** [reply]
> migrate homebrew-cask-fonts to homebrew-fonts which can be shared between macOS and Linux and managed in the Homebrew organisation and provide a brew fonts to do any of the symlinking required on macOS

This outcome seems the best possible if @tani is volunteering to do the work and maintenance/support doesn't overly burden the Cask maintainers. We could allowlist `homebrew/fonts` as the only cask tap supported on Linux.

**Reply by carlocab** [reply]
`Homebrew/cask-fonts` doesn't require too much maintenance, and relies very heavily on automation (more so than any other Homebrew repository). That also seems to be @tani's approach to `linuxbrew/fonts` so it seems like this could work well.

**Reply by vitorgalvao** [reply]
> `Homebrew/cask-fonts` doesn't require too much maintenance, and relies very heavily on automation (more so than any other Homebrew repository).

**Because** it used to require so much maintenance and was so frequently outdated that the way to make it tenable was to rely on the work of others who are more invested (Google Fonts[^1]) and [reject everything else](https://github.com/Homebrew/homebrew-cask-fonts/blob/master/CONTRIBUTING.md#when-to-contribute).

In other words, that repo is low maintenance because no one wants to maintain it. If you’re going to expand it, my advice is to ensure someone has the will and availability to take it to completion, lest we end up in a situation where it again needs regular maintenance and everything falls apart.

[^1]: In its current form, if Google abandons Google Fonts the repo dies on the spot.

**Reply by carlocab** [reply]
I don't think anyone wants to expand Cask fonts; just make it available to Linux users.

**Reply by vitorgalvao** [reply]
By expanding, what I mean is that this might be one of those cases where the last 10% require 90% of the effort. For example, there needs to be a decision on how to deal with PKG fonts. Should they be removed (users won’t like that, especially for Apple fonts), hidden on Linux, something else? What about other edge cases, do they exist?

I’m not saying “don’t do it”. By all means do, I like the idea of fonts on Linux. I’m only urging you to keep in mind that historically there have very few people willing to take up the slack from that repo, and that if something is abandoned half-way it may stay like that. There’s a reason I keep insisting that repo is more trouble than it’s worth.

**Reply by carlocab** [reply]
> For example, there needs to be a decision on how to deal with PKG fonts.

There are about five of those left. (If I'm counting correctly...) I'm not hugely bothered either way about what to do about them.

> What about other edge cases, do they exist?

Maybe; but Homebrew/core is full of cases like these, and Linux users don't seem too unhappy about them. By "cases like these" I mean formulae that have macOS-specific assumptions baked in that make them a little bit (or even a lot) broken on Linux.

Of course, there's a lot more interest in Homebrew/core so it's probably overall better maintained (maybe? It's hard to compare.). But my point is that I think edge cases probably aren't all that troublesome. (Or, at least, we ought not to worry too much about these.) The existence of linuxbrew/fonts probably also supports this.

> There’s a reason I keep insisting that repo is more trouble than it’s worth.

Maybe we can interest the owner of this repo. https://github.com/alexmyczko/fnt

**Reply by vitorgalvao** [reply]
All good points, as I’ve come to expect from you! Nothing to add.

**Reply by tani** [reply]
> This outcome seems the best possible if @tani is volunteering to do the work and maintenance/support doesn't overly burden the Cask maintainers. We could allowlist homebrew/fonts as the only cask tap supported on Linux.

Of course, I would like to be volunteering to do the work. Shall we cooperate macOS/ Linux support together? As @vitorgalvao said, there is not so much I can contribute because it is working with Google Fonts system to generate the Casks. I will work to remove some edge cases for the Casks. In fact, I have dropped some fonts, e.g., sfmono, to port homebrew-cask-fonts to Linux for Linuxbrew/fonts.

**Reply by MikeMcQuaid** [reply]
> Per @colindean on Slack, it seems that making font casks work on Linux requires changing only a handful of lines of code.

Changing the code? Sure. Handling all the edge-cases and the conceptual difference that "casks don't work on Linux (except for these ones that actually do): much harder and I'm really not convinced it's worth it.

> * migrate homebrew-cask-fonts formulae to tani's personal GitHub account for Linux-only

Another option: have a homebrew-linux-fonts or something tap moved into the Homebrew organisation and we direct folks on Linux to tap that instead and don't allow macOS folks to tap it.

**Reply by carlocab** [reply]
Edge cases I'm not too worried about. The conceptual difference, I am. I think this can be handled by code, too, though: instead of calling them `cask`s just make them `font`s, where `font` is actually just (privately) an alias for `cask`. That said:

> Another option: have a homebrew-linux-fonts or something tap moved into the Homebrew organisation and we direct folks on Linux to tap that instead and don't allow macOS folks to tap it.

I'm fine with this. It's simpler too.

**Reply by tani** [reply]
> Another option: have a homebrew-linux-fonts or something tap moved into the Homebrew organisation and we direct folks on Linux to tap that instead and don't allow macOS folks to tap it.

This is the same as the current situation, isn't this? We have `linuxbrew/homebrew-fonts`.
It seems there is no problem if the problem is only the security of the organization.

**Reply by MikeMcQuaid** [reply]
More or less but I think we should move it into the Homebrew organisation.

**Reply by tani** [reply]
Then, shall we close this discussion with the following steps?

1. Transfer `linuxbew/homebrew-fonts` to `homebrew/homebrew-linux-fonts`.
2. Assign the repository administration account to @tani to keep maintaining the repository.
3. Open the new issue if we want to merge this project to macOS's casks, `homebrew-cask-fonts` in the repository.

**Reply by MikeMcQuaid** [reply]
> Transfer linuxbew/homebrew-fonts to homebrew/homebrew-linux-fonts.
> Assign the repository administration account to @tani to keep maintaining the repository.

Sorry for the (very) late response here @tani. 

Any objections to this @carlocab @colindean @vitorgalvao @sjackman @Rylan12?

**Reply by colindean** [reply]
I'm okay with this.

**Reply by tani** [reply]
Thank you for the reactions. Then, could I keep maintaining formulae in `homebrew/homebrew-linux-fonts`?

**Reply by carlocab** [reply]
Sounds good!

**Reply by MikeMcQuaid** [reply]
Done: https://github.com/Homebrew/homebrew-linux-fonts!

Thanks all.

**Reply by tani** [reply]
Thank you so much. I am happy to contribute to Homebrew project!

**Reply by carlocab** [reply]
Thanks for all your work, @tani!

**Reply by colindean** [reply]
I'm very excited for this!
